test/alternator/test_cql_rbac.py

```diff
@@ -0,0 +1,217 @@
+
+import pytest
+import boto3
+from botocore.exceptions import ClientError
+import time
+from contextlib import contextmanager
+from test.alternator.util import is_aws, unique_table_name, random_string
+
+from cassandra.auth import PlainTextAuthProvider
+from cassandra.cluster import Cluster, ConsistencyLevel, ExecutionProfile, EXEC_PROFILE_DEFAULT
+from cassandra.policies import RoundRobinPolicy
+import re
+
+# This file is all about testing RBAC as configured via CQL, so we need to
+# connect to CQL to set these tests up. The "cql" fixture below enables that.
+# If we're not testing Scylla, or the CQL port is not available on the same
+# IP address as the Alternator IP address, a test using this fixture will
+# be skipped with a message about the CQL API not being available.
+@pytest.fixture(scope="module")
+def cql(dynamodb):
+    if is_aws(dynamodb):
+        pytest.skip('Scylla-only CQL API not supported by AWS')
+    url = dynamodb.meta.client._endpoint.host
+    host, = re.search(r'.*://([^:]*):', url).groups()
+    profile = ExecutionProfile(
+        load_balancing_policy=RoundRobinPolicy(),
+        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
+        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL)
+    cluster = Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile},
+        contact_points=[host],
+        port=9042,
+        protocol_version=4,
+        auth_provider=PlainTextAuthProvider(username='cassandra', password='cassandra'),
+    )
+    try:
+        ret = cluster.connect()
+        # "BEGIN BATCH APPLY BATCH" is the closest to do-nothing I could find
+        ret.execute("BEGIN BATCH APPLY BATCH")
+    except:
+        pytest.skip('Could not connect to Scylla-only CQL API')
+    yield ret
+    cluster.shutdown()
+
+# new_role() is a context manager for temporarily creating a new role with
+# a unique name and returning its name and the secret key needed to connect
+# to it with the DynamoDB API.
+# The "login" and "superuser" flags are passed to the CREATE ROLE statement.
+@contextmanager
+def new_role(cql, login=True, superuser=False):
+    # The role name is not a table's name but it doesn't matter. Because our
+    # unique_table_name() uses (deliberately) a non-lower-case character, the
+    # role name has to be quoted in double quotes when used in CQL below.
+    role = unique_table_name()
+    # The password set for the new role is identical to the user name (not
+    # very secure ;-)) - but we later need to retrieve the "salted hash" of
+    # this password, which serves in Alternator as the secret key of the role.
+    cql.execute(f"CREATE ROLE \"{role}\" WITH PASSWORD = '{role}' AND SUPERUSER = {superuser} AND LOGIN = {login}")
+    # Newer Scylla places the "roles" table in the "system" keyspace, but
+    # older versions used "system_auth_v2" or "system_auth"
+    key = None
+    for ks in ['system', 'system_auth_v2', 'system_auth']:
```

Comment:
Q: why is system_auth interesting here?

Reply:
It's interesting if you want to run the new test on an older branch of Scylla. If I ever do that (e.g., for checking missing backports), I want to see tests fail on real bugs (like not handling login=FALSE) - not on test implementation details.

Even more importantly, it is theoretically possible that we'll want to backport a patch (and its test) to an older version of Scylla which still uses one of these older keyspaces. If we really want (and I hope we don't want), the new Alternator RBAC can be backported to an older version of Scylla which still uses the old name `system_auth.roles`.

I agree it's not super-important, but since I already wrote this code snippet in the past for other reasons (some CQL test), it was easy for me to copy it as-is and not "simplify" it.

Reply:
> If we really want (and I hope we don't want), the new Alternator RBAC can be backported to an older version of Scylla which still uses the old name system_auth.roles.

I'm trying to establish which version of scylla our client is using and potentially to how old release we'll have to backport this feature + how exactly we'll deploy it. I'm afraid that whatever we agree on now in this regard may change in the near future, so better keep in mind the worst case scenario, i.e. authz will have to be backported to pre-raft authn.

Reply:
I doubt that (pre-raft), but we should not close the door on this if we don't have to / make our life easier.

Reply:
So the client is using the latest version and will most probably be OK with upgrading to the newest version that has this feature, so most probably we don't have to worry about backports.

Reply:
I still want, if you don't strongly oppose this, to allow my test to run also against older versions of Scylla if it's not too difficult. In case a year from now I'm asked to check if some random version of Scylla has this feature or not, and I want to run this test against it.

Reply:
I can only support this idea, also because it doesn't sound like a big effort.